File: surface/widgets/winui/components/widgets/primitives/Presence.cs
// ============================================================
// Clef Surface WinUI Widget — Presence
//
// Controls conditional rendering of children based on the
// present flag. Supports animated enter/exit transitions via
// WinUI 3 implicit animations. ForceMount keeps content in
// the visual tree even when not visible.
//
// Adapts the presence.widget spec: anatomy (root), states
// (unmounted, mounting, mounted, unmounting), and connect
// attributes (data-part, data-state, data-present) to WinUI 3.
// ============================================================

using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;

namespace Clef.Surface.WinUI.Widgets.Primitives;

public sealed class ClefPresence : UserControl
{
    public static readonly DependencyProperty IsPresentProperty =
        DependencyProperty.Register(nameof(IsPresent), typeof(bool), typeof(ClefPresence),
            new PropertyMetadata(false, OnPropertyChanged));

    public static readonly DependencyProperty ForceMountProperty =
        DependencyProperty.Register(nameof(ForceMount), typeof(bool), typeof(ClefPresence),
            new PropertyMetadata(false, OnPropertyChanged));

    public bool IsPresent { get => (bool)GetValue(IsPresentProperty); set => SetValue(IsPresentProperty, value); }
    public bool ForceMount { get => (bool)GetValue(ForceMountProperty); set => SetValue(ForceMountProperty, value); }

    private readonly ContentPresenter _presenter;

    public ClefPresence()
    {
        _presenter = new ContentPresenter();
        Content = _presenter;
        UpdateVisual();
    }

    public void SetPresenceContent(UIElement element)
    {
        _presenter.Content = element;
        UpdateVisual();
    }

    private static void OnPropertyChanged(DependencyObject d, DependencyPropertyChangedEventArgs e)
        => ((ClefPresence)d).UpdateVisual();

    private void UpdateVisual()
    {
        if (ForceMount)
        {
            _presenter.Visibility = Visibility.Visible;
            _presenter.Opacity = IsPresent ? 1.0 : 0.0;
        }
        else
        {
            _presenter.Visibility = IsPresent ? Visibility.Visible : Visibility.Collapsed;
            _presenter.Opacity = 1.0;
        }
    }
}

File: surface/widgets/gtk/components/widgets/primitives/Presence.ts
// ============================================================
// Clef Surface GTK Widget — Presence
//
// Conditional rendering wrapper with animated enter/exit
// transitions. Uses Gtk.Revealer for expand/collapse
// animations when the child mounts or unmounts.
//
// Adapts the presence.widget spec: anatomy (root), states
// (present, hidden), and connect attributes to GTK4 rendering.
// ============================================================

import Gtk from 'gi://Gtk?version=4.0';

// --------------- Props ---------------

export interface PresenceProps {
  present?: boolean;
  child?: Gtk.Widget | null;
  transitionType?: Gtk.RevealerTransitionType;
  transitionDuration?: number;
}

// --------------- Component ---------------

/**
 * Creates a GTK4 Revealer for conditional rendering with
 * animated enter/exit transitions.
 */
export function createPresence(props: PresenceProps = {}): Gtk.Widget {
  const {
    present = true,
    child = null,
    transitionType = Gtk.RevealerTransitionType.CROSSFADE,
    transitionDuration = 200,
  } = props;

  const revealer = new Gtk.Revealer({
    revealChild: present,
    transitionType,
    transitionDuration,
  });

  if (child) {
    revealer.set_child(child);
  }

  return revealer;
}
